dev/provider_packages/prepare_provider_packages.py

```diff
@@ -314,7 +314,20 @@ def get_install_requirements(provider_package_id: str, version_suffix: str) -> s
 
     :return: install requirements of the package
     """
-    install_requires = ALL_DEPENDENCIES[provider_package_id][DEPS]
+
+    def apply_version_suffix(install_clause: str) -> str:
+        if install_clause.startswith("apache-airflow") and ">=" in install_clause and version_suffix != "":
+            # This is workaround for `pip` bug. When you specify dependency as >= X.Y.Z, and you
+            # Have packages X.Y.Zdev0 or X.Y.Zrc1, the "=" part does not include the pre-release versions
+            # you need to explicitly specify >= X.Y.Za0 to include pre-release versions.
+            # Therefore, when we are preparing pre-releases we always add a0 for airflow and other providers
+            # to Allow installing two packages released in rc* that depend on each-other
```

Comment:
This is not a bug but in the Python version range specification. A better “fix” would be to conditionally supply `--pre` when pip is invoked.

Reply:
Ah. I thought you mentioned before that this was a long standing bug that you never got around to fix, but I must have been mistaken. If -pre works this way, then it is indeed a much better fix, so let me change it

Reply:
Sorry - i was 100% sure we discussed it before and that was the outcome :)

Reply:
I tried to use pre (I will have a nice feature added to docker files out of that), but I am afraid this is indeed the bug and I remembered our earlier discussion well. This is what happened when I tried to install the providers:

```
pip install --find-links=file:///docker-context-files 
--root-user-action ignore --upgrade --upgrade-strategy eager 
--pre 'apache-airflow[amazon,async,celery,cncf.kubernetes,dask,docker,elasticsearch,ftp,google,google_auth,grpc,hashicorp,http,ldap,microsoft.azure,mysql,odbc,pandas,postgres,redis,sendgrid,sftp,slack,ssh,statsd,virtualenv]==2.4.0.dev0'
 /docker-context-files/apache_airflow_providers_amazon-4.1.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_celery-3.0.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_cncf_kubernetes-4.2.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_common_sql-1.1.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_docker-3.1.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_elasticsearch-4.1.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_ftp-3.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_google-8.2.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_grpc-3.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_hashicorp-3.0.1.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_http-4.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_imap-3.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_microsoft_azure-4.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_mysql-3.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_odbc-3.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_postgres-5.1.0.dev0-py3-none-any.whl 
/docker-context-files/apache_airflow_providers_redis-3.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_sendgrid-3.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_sftp-4.0.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_slack-5.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_sqlite-3.1.0.dev0-py3-none-any.whl
/docker-context-files/apache_airflow_providers_ssh-3.1.0.dev0-py3-none-any.whl 'dill<0.3.3'
```

This fails with: 

```
#56 34.54 ERROR: Cannot install apache-airflow-providers-amazon==4.1.0.dev0, apache-airflow-providers-common-sql 1.1.0.dev0 (from /docker-context-files/apache_airflow_providers_common_sql-1.1.0.dev0-py3-none-any.whl), apache-airflow-providers-elasticsearch==4.1.0.dev0, apache-airflow-providers-google==8.2.0.dev0 and apache-airflow[amazon,async,celery,cncf-kubernetes,dask,docker,elasticsearch,ftp,google,google-auth,grpc,hashicorp,http,ldap,microsoft-azure,mysql,odbc,pandas,postgres,redis,sendgrid,sftp,slack,ssh,statsd,virtualenv]==2.4.0.dev0 because these package versions have conflicting dependencies.
#56 34.54 
#56 34.54 The conflict is caused by:
#56 34.54     The user requested apache-airflow-providers-common-sql 1.1.0.dev0 (from /docker-context-files/apache_airflow_providers_common_sql-1.1.0.dev0-py3-none-any.whl)
#56 34.54     apache-airflow[amazon,async,celery,cncf-kubernetes,dask,docker,elasticsearch,ftp,google,google-auth,grpc,hashicorp,http,ldap,microsoft-azure,mysql,odbc,pandas,postgres,redis,sendgrid,sftp,slack,ssh,statsd,virtualenv] 2.4.0.dev0 depends on apache-airflow-providers-common-sql
#56 34.54     apache-airflow-providers-amazon 4.1.0.dev0 depends on apache-airflow-providers-common-sql
#56 34.54     apache-airflow-providers-elasticsearch 4.1.0.dev0 depends on apache-airflow-providers-common-sql
#56 34.54     apache-airflow-providers-google 8.2.0.dev0 depends on apache-airflow-providers-common-sql>=1.1.0
#56 34.54 
#56 34.54 To fix this you could try to:
#56 34.54 1. loosen the range of package versions you've specified
#56 34.54 2. remove package versions to allow pip attempt to solve the dependency conflict
#56 34.54 
#56 34.54 ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

The problem is that 

* apache-airflow-providers-google 8.2.0.dev0 depends on apache-airflow-providers-common-sql>=1.1.0
* The user requested apache-airflow-providers-common-sql 1.1.0.dev0 (from /docker-context-files/apache_airflow_providers_common_sql-1.1.0.dev0-py3-none-any.whl

I was using ``--pre`` flag (as you can see above). 

Still google's > 1.1.0 conflicts with 1.1.0dev0 - which is the problem

Reply:
cc: @uranusjr - somethng I missed?

Reply:
BTW. The possibility of adding `--pre` flag (or any other) added here: https://github.com/apache/airflow/pull/25337 (but as above, it does not seem to solve the problem).

Reply:
The only way I could make it work @uranusjr was to remove the limits altogether. I am not sure if there is another way ?

Reply:
The current  version (with limit removal) is also running in #24836  to double check if it works (but I tested it locally and it worked).

Reply:
I think the problem is that 1.1.0 prereleases are earlier 1.1.0 (because prereleases are logically before the stable one). Adding `dev0` instead would probably work (and is probably an appropriate solution for Airflow in this use case, since we are kind of intentionally breaking the version range for testing).

Reply:
Yeah, I will experiment a bit with it. I do not like to remove the limits altogether. Still it's a bit surprising it did not work when I put `>=1.1.0.a0` -> I would imagine ".dev0" is higher than ".a0".

BTW. This is a bit experimental with `common.sql`, in the way that we have a chance now to test how this case works - but we need to solve it 'well' for the future - especially if we seriously think about splitting google provider  - this will introduce even more complex relations as there will likely be a 'google.common' or even 'alphabet,gcp.common` provider that other google providers will depend on and it will be quite often that we will have to released those together. 

And we need to give a chance for our users to be able to test rc's when we release them as well.

I am not sure if that one is because we are using wheel packages from local folders only? Will that work the same when we release rc's in PyPI?. That will make it a bit harder to test if our - in this case - `google8.1.0rc1 has >= common.sql.1.1.0rc1 requirement" - will the user get the rc1 of common sql installed (even if using pre?).

Reply:
`dev0` not higher than `a0` (or lower in the same sense) because it’s actually a separate segment altogether. In PEP 440, a `dev0` part belongs to the development segment, while `a0` belongs to the pre-release segment. So adding `a0` to the range specifier does not help identifying dev releases, but `dev0` can.

https://peps.python.org/pep-0440/#pre-releases

Reply:
Yeah. That woudl explain it. Just checked it and indeed - simply replacing the suffix with the one we are just building with is good. Update coming in a minute.

So really the problem (might not be a "bug" ) with `pip` is that `--pre` does not apply to limit of the dependecies of the packages you are installing - just to the packages that you are installing. Would that be the right "assesment" ?

Do you think is it something that could be changed easiy and accepted? Maybe I could even attempt to find and fix it, but I am a little bit sceptical after the rejections (and the form of those rejections especially) I got before, so I am a little bit sceptical about submitting anything like that without knowing if it is something that maybe has been discussed in the past or maybe there is a long discussion somewhere I am not aware of.

Reply:
I think `--pre` does apply to transitive dependencies, but it’s not possible to tell without a reproducible with much less moving parts.

Reply:
Or is `--pre` only valid for pre-releases and we need `--dev` as well? (Just reading PEP 440 and it seems to me that `--pre` would only apply to a/b ones but not dev ones ?

Reply:
Though the other part of PEP mentions ("Pre-releases of any kind, including developmental releases") - so I am not sure any more whether `--pre`'s pre-release also includes dev or not :)

Reply:
Actually after further reading of the PEP, I tested one more approach. It's not explicitly mentioned in the PEP, but apparently it works.

Seems that converting the "pre-release" depencies to have  ".*" at the end even in >= would include all kinds of pre-releases, so at least we have a way to specify "all pre-releases including development ones" in our dependencies. So `apache-airflow-providers-common-sql>=1.1.0.*` works in this case and that's what I am going to settle with I think. 

This is the part of meta-data fron .whl in the Google Provider that works:

```
Requires-Dist: PyOpenSSL
Requires-Dist: apache-airflow-providers-common-sql (>=1.1.0.*)
Requires-Dist: apache-airflow (>=2.2.0.*)
Requires-Dist: google-ads (>=15.1.1)
```

I think that's sustainable for our case (but we have to manually modify the requirements). Would be great to add this behaviour to `--pre` flag I guess, but we can definitely live with that even if it is not added.